File: languages/rust/src/polar.rs
/// Communicate with the Polar virtual machine: load rules, make queries, etc/
use std::fs::File;
use std::io::Read;
use std::rc::Rc;
use std::sync::{Arc, Mutex};

use crate::host::Host;

pub struct Polar {
    inner: Rc<crate::PolarCore>,
    host: Arc<Mutex<Host>>,
}

impl Polar {
    pub fn new() -> Self {
        let inner = Rc::new(crate::PolarCore::new());
        let mut host = Host::new(Rc::downgrade(&inner));

        // register all builtin constants
        for (name, class, value) in crate::host::builtins::constants(&mut host) {
            host.cache_class(class, Some(name.clone()));
            inner.register_constant(name, value);
        }

        Self {
            host: Arc::new(Mutex::new(host)),
            inner,
        }
    }

    pub fn clear(&mut self) {
        *self = Self::new();
    }

    fn check_inline_queries(&mut self) -> anyhow::Result<()> {
        while let Some(q) = self.inner.next_inline_query(false) {
            let query = crate::query::Query::new(q, self.host.clone());
            match query.collect::<anyhow::Result<Vec<_>>>() {
                Ok(v) if !v.is_empty() => continue,
                Ok(_) => anyhow::bail!("inline query result was false"),
                Err(e) => {
                    anyhow::bail!("error in inline query: {}", e);
                }
            }
        }
        Ok(())
    }

    pub fn load_file(&mut self, file: &str) -> anyhow::Result<()> {
        let mut f = File::open(&file)?;
        let mut policy = String::new();
        f.read_to_string(&mut policy)?;
        self.inner.load(&policy, Some(file.to_string()))?;

        self.check_inline_queries()
    }

    pub fn load_str(&mut self, s: &str) -> anyhow::Result<()> {
        self.inner.load(s, None)?;
        self.check_inline_queries()
    }

    pub fn query(&mut self, s: &str) -> anyhow::Result<crate::query::Query> {
        let query = self.inner.new_query(s, false)?;
        let query = crate::query::Query::new(query, self.host.clone());
        Ok(query)
    }

    pub fn register_class(&mut self, class: crate::host::Class) -> anyhow::Result<()> {
        let mut host = self.host.lock().unwrap();
        let name = class.name.clone();
        let _name = host.cache_class(class, Some(polar_core::types::Symbol(name)));
        Ok(())
    }

    pub fn register_constant<V: crate::host::ToPolar>(
        &mut self,
        name: &str,
        value: &V,
    ) -> anyhow::Result<()> {
        let mut host = self.host.lock().unwrap();
        self.inner.register_constant(
            polar_core::types::Symbol(name.to_string()),
            host.to_polar(value),
        );
        Ok(())
    }

    pub fn query_rule<'a>(
        &mut self,
        name: &str,
        args: impl IntoIterator<Item = &'a dyn crate::host::ToPolar>,
    ) -> anyhow::Result<crate::query::Query> {
        let args = args
            .into_iter()
            .map(|arg| arg.to_polar(&mut self.host.lock().unwrap()))
            .collect();
        let query_value = polar_core::types::Value::Call(polar_core::types::Predicate {
            name: polar_core::types::Symbol(name.to_string()),
            args,
        });
        let query_term = polar_core::types::Term::new_from_ffi(query_value);
        let query = self.inner.new_query_from_term(query_term, false);
        let query = crate::query::Query::new(query, self.host.clone());
        Ok(query)
    }
}

File: languages/go/pkg/polar.go
// """Communicate with the Polar virtual machine: load rules, make queries, etc."""

package oso

import (
	"fmt"
	"io/ioutil"
	"path/filepath"
	"reflect"
)

var CLASSES = make(map[string]reflect.Type)

type Polar struct {
	ffiPolar PolarFfi
	host     Host
}

func NewPolar() Polar {
	ffiPolar := NewPolarFfi()
	polar := Polar{
		ffiPolar: ffiPolar,
		host:     NewHost(ffiPolar),
	}

	// register global constants
	return polar
}

// CLASSES: Dict[str, type] = {}

// class Polar:
//     """Polar API"""

//     def __init__(self, classes=CLASSES):
//         self.ffi_polar = FfiPolar()
//         self.host = Host(self.ffi_polar)

//         # Register global constants.
//         self.register_constant(None, name="nil")

//         # Register built-in classes.
//         self.register_class(bool, name="Boolean")
//         self.register_class(int, name="Integer")
//         self.register_class(float, name="Float")
//         self.register_class(list, name="List")
//         self.register_class(dict, name="Dictionary")
//         self.register_class(str, name="String")
//         self.register_class(datetime, name="Datetime")
//         self.register_class(timedelta, name="Timedelta")

//         # Pre-registered classes.
//         for name, cls in classes.items():
//             self.register_class(cls, name=name)

//     def __del__(self):
//         del self.host
//         del self.ffi_polar

func (p Polar) checkInlineQueries() error {
	for {
		query, err := p.ffiPolar.nextInlineQuery()
		if err != nil {
			return err
		}
		if query == nil {
			return nil
		}
		// TODO
		// try:
		// 	next(Query(query, host=self.host.copy()).run())
		// except StopIteration:
		// 	source = query.source()
		// 	raise InlineQueryFailedError(source.get())
	}

	return nil
}

func (p Polar) LoadFile(f string) error {
	if filepath.Ext(f) != "polar" {
		return &PolarFileExtensionError{file: f}
	}

	data, err := ioutil.ReadFile(f)
	if err != nil {
		return err
	}
	err = p.ffiPolar.load(string(data), &f)
	if err != nil {
		return err
	}
	return p.checkInlineQueries()
}

func (p Polar) LoadString(s string) error {
	err := p.ffiPolar.load(s, nil)
	if err != nil {
		return err
	}
	return p.checkInlineQueries()
}

func (p Polar) ClearRules() error {
	return p.ffiPolar.clearRules()
}

// Query accepts string and predicates
func (p Polar) Query(query interface{}) (*Query, error) {
	switch query.(type) {
	case string:
		ffiQuery, err := p.ffiPolar.newQueryFromStr(query.(string))
		if err != nil {
			return nil, err
		}
		newQuery := newQuery(*ffiQuery, p.host.copy())
		return &newQuery, nil
	}

	return nil, fmt.Errorf("Unsupported query type: %v", query)
}

func (p Polar) QueryRule(name string, args ...interface{}) (*Query, error) {
	// TODO
	return nil, fmt.Errorf("Unsupported: QueryRule")
}

func (p Polar) Repl(files ...string) error {
	return fmt.Errorf("Go REPL is not yet implemented")
}

//     def repl(self, files=[]):
//         """Start an interactive REPL session."""
//         for f in files:
//             self.load_file(f)

//         while True:
//             try:
//                 query = input(FG_BLUE + "query> " + RESET).strip(";")
//             except (EOFError, KeyboardInterrupt):
//                 return
//             try:
//                 ffi_query = self.ffi_polar.new_query_from_str(query)
//             except ParserError as e:
//                 print_error(e)
//                 continue

//             result = False
//             try:
//                 query = Query(ffi_query, host=self.host.copy()).run()
//                 for res in query:
//                     result = True
//                     bindings = res["bindings"]
//                     if bindings:
//                         for variable, value in bindings.items():
//                             print(variable + " = " + repr(value))
//                     else:
//                         print(True)
//             except PolarRuntimeError as e:
//                 print_error(e)
//                 continue
//             if not result:
//                 print(False)

func (p Polar) RegisterClass(cls reflect.Type, name *string) error {
	err := p.host.cacheClass(cls, name)
	if err != nil {
		return err
	}
	return p.RegisterConstant(cls, cls.Name())
}

func (p Polar) RegisterConstant(value interface{}, name string) error {
	polarValue, err := p.host.toPolar(value)
	if err != nil {
		return err
	}
	return p.ffiPolar.registerConstant(*polarValue, name)
}
